rfcs/0112-demote-darwin-support-tier.md

```diff
@@ -0,0 +1,91 @@
+
+An excerpt of these issues are:
+
+- The only way to debug issues without owning a Mac is to run Darwin in a VM, of which the legal situation is unclear.
+   - Those licensing requirements are causing friction with our CI setup as well.
+- The sandboxing capabilities on MacOS are limited. This results in a special category of test failures only relevant to Darwin. It also prevents us from providing a community builder, because it could not be shared between multiple people that cannot trust each other.
+- The MacOS SDK cannot be updated and is stuck on 10.12 (released 2016/09) because Apple does not publish the required sources.
+    - https://github.com/NixOS/nixpkgs/issues/101229#issuecomment-938747052
+    - It is highly unlikely that we can report bugs to projects with such an outdated and unsupported SDK version
+- The MacOS SDK being outdated blocks nixpkgs moving to go 1.17:
+    - https://github.com/NixOS/nixpkgs/pull/127519#issuecomment-864926149
+    - This will become a growing pain point once more packages become go 1.17-only (tailscale, talosctl, …)
+- Multiple Packages were marked as broken, because they require symbols from newer MacOS SDK versions.
+    - https://github.com/NixOS/nixpkgs/commit/3ceb5ab5ed0d1fcbe53ef00621f44d61dc524796
+    - https://github.com/NixOS/nixpkgs/commit/c9a3ac5d3cb5e910238d01e534d74d5d50e4b6b7
+- Curl added a SystemConfiguration dependency for NAT64 support, which introduced a reference loop, requiring a downstream patch to workaround.
+    - This blocked updating to more recent curl versions for most of the last release cycle.
+    - https://github.com/NixOS/nixpkgs/pull/124502#issuecomment-850834981
+- Enabling brotli support by default in curl broke the Darwin stdenv, which draws in a great number of packages.
+    - https://github.com/NixOS/nixpkgs/pull/112947
+    - https://github.com/NixOS/nixpkgs/pull/115498
+
+# Detailed design
+[design]: #detailed-design
+
+- Together with this RFC, `0046-platform-support-tiers.md` will be updated accordingly (simply move `x86_64-darwin` down one section).
+- Whatever needs to be done to make Darwin not block any channels anymore (TODO).
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+- Failing tests on `x86_64-darwin` will not block channel advances in case of failures anymore.
+- The existing testing infrastructure will remain unchanged. Tests will continue to be run to the extent the infrastructure can provide.
+
+# Drawbacks
+[drawbacks]: #drawbacks
```

Comment:
This might cause a bit of a feedback loop:

1. More darwin breakages introduced.
2. Bigger gap between capacity of darwin contributors and capacity requirements of darwin breakages.
3. Quality of darwin support lowers.
4. Influx of new darwin users decreases / outflux of darwin users increases.
5. Influx of new darwin contributors decreases / outflux of darwin contributors increases.
6. Go to 2.

Reply:
Darwin is currently the only non-linux platform currently supported in tier 1 and 2. It is likely that by demoting it support for all non-linux platforms will suffer. That is partly by design: a lot of the pain points raised in this RFC are related to multiplatform development workflows, so having to support only a single platform will improve all those points, but it's worth highlighting this consequence explicitly.

Reply:
> This might cause a bit of a feedback loop

I agree with the first thee points, but I expect that the difference for end users will be small enough that 4. won't happen. Especially because Nix on Darwin is always a gradual opt-in instead of being a total buy-in like NixOS.

> It is likely that by demoting it support for all non-linux platforms will suffer.

Which non-Linux platforms do you have in mind and what does Darwin support do for them? Scrolling through the RFC 46 platform list, I see some embedded and `mingw` targets in Tier 4 which are hardly related to Darwin, and all others are unsupported (Tier 7).

Reply:
NetBSD situation has apparently improved since, and I guess FreeBSD/NetBSD support would benefit from better clang support, but mayb Firefox + musl-Linux would force us to keep up anyway?

Reply:
> Which non-Linux platforms do you have in mind and what does Darwin support do for them?

I checked the supported platforms and I was surprised at finding out what they were, but I didn't have any specific platform in mind, my point is not that darwin support would be helpful to other platforms because it has something in common with them, it's that it stops us from assuming linux is the only platform we need to support.

In particular note this part of rfc 46:
> Tier 3:
> Completely platform-specific fixes are expected to be rare and non-intrusive.

If we consider only linux will be in tiers 1 and 2, that might result in any fix requiring a big change because the existing code assumes only linux to be rejected as too intrusive.

Maybe supporting only linux is what we'd prefer anyway, but it would be better to make that explicit then.

Reply:
The word «completely» is meaningful there — the next paragraph reads 

> General cleanups of non-standard assumptions (e.g. «everything that is not x86 is a kind of ARM» or «malloc(0) behaviour is a reliable indicator of other malloc features») useful for these platforms are welcome.

So currently-musl-only fixes have been justified before as «unconditionally assuming this or that is a bug, and enough of musl work has been shown usable that our fix should be accepted».

Specifically Darwin is of course more at risk, as often needs source code patches (not just extra inputs or whatever) that are specific not even simply to macOS but also to which trade-offs have to be made for Nixpkgs to work on macOS.